File: src/main/java/org/tindalos/principle/app/service/InputValidator.scala
package org.tindalos.principle.app.service

import org.tindalos.principle.domain.core.AnalysisPlan

/**
 * Created by mate.magyari on 03/08/2014.
 */
object InputValidator {

  def validate(plan: AnalysisPlan) = {

    if (plan.expectations.thirdParty == null) (true, "")
    else {
      val layers = plan.expectations.layering.layers
      val barriers = plan.expectations.thirdParty.barriers

      val invalidBarriers = for (barrier <- barriers
                                 if (!layers.contains(barrier.layer))) yield barrier



      if (!invalidBarriers.isEmpty)
        (false, s"Invalid layers specified under Barriers: ${invalidBarriers}")
      else {
        def layerOrder(a: String, b: String) = layers.indexOf(a) < layers.indexOf(b)
        val layersOfBarriers = barriers.map(_.layer)
        if (!layersOfBarriers.sortWith(layerOrder).equals(layersOfBarriers))
          (false, "The order of layers in barriers should be the same as of under layering")
        else (true, "")
      }
    }
  }

}

File: src/main/java/org/tindalos/principle/app/service/InputValidator.java
package org.tindalos.principle.app.service;

import org.tindalos.principle.domain.core.AnalysisPlan;

/**
 * Validates analysis plan inputs to ensure consistency and correctness
 * of configuration before running architectural analysis.
 */
public interface InputValidator {

    /**
     * Validates the analysis plan configuration.
     *
     * @param plan the analysis plan to validate
     * @return validation result indicating success or failure with error message
     */
    ValidationResult validate(AnalysisPlan plan);

}

